File: 2019/J5.cpp
//works but is slower than python so currently in progress

#define _CRT_SECURE_NO_WARNINGS
#include <bits/stdc++.h>
#include <unordered_map>
#include <unordered_set>
using namespace std;

vector<int> rabin_karp(string word, string text)
{
	vector<int> returned;
	int wordlength = word.length();
	int textlength = text.length();
	int whash = 0;
	int thash = 0;
	int prime = 7;
	int inputcharcount = 2;
	int i;
	int j;
	int hash = pow(inputcharcount, wordlength - 1);
	hash %= prime;

	for (i = 0; i < wordlength; i++)
	{
		whash = (inputcharcount * whash + word[i]) % prime;
		thash = (inputcharcount * thash + text[i]) % prime;
	}

	for (i = 0; i <= textlength - wordlength; i++)
	{
		if (whash == thash)
		{
			for (j = 0; j < wordlength; j++)
			{
				if (text[i + j] != word[j])
					break;
			}
			if(j == wordlength)
			{
				returned.push_back(i);
			}
		}

		if (i < textlength - wordlength)
		{
			thash = inputcharcount * (thash - text[i] * hash) + text[i + wordlength];
			thash %= prime;

			if (thash < 0)
			{
				thash = (thash + prime);
			}
		}
	}
	return returned;
}

bool found = false;
unordered_map<string, vector<string>> memoize;
//unordered_map<string, unordered_map<string, vector<string>>> memoizeparams;
bool loop(string currentstring, string endstring, vector<string> rulesnum, vector<int> indexes, vector<string> past, int iteration, int maxiteration, unordered_map<string, vector<string>> rules)
{
	if (!found)
	{
		if (iteration == maxiteration)
		{
			if (currentstring == endstring)
			{
				for (int i = 0; i < maxiteration; i++)
				{
					cout << rulesnum[i] << " " << indexes[i] << " "<< past[i] << "\n";
				}
				found = true;
				return true;
			}
			else
			{
				return false;
			}
		}
		else
		{
			if (memoize.find(currentstring) == memoize.end())
			{
				memoize[currentstring] = {};
				//memoizeparams[currentstring] = {};
				for (pair<string, vector<string>> rule : rules)
				{
					vector<int> found = rabin_karp(rule.first, currentstring);
					if (!found.empty())
					{
						for (int index : found)
						{
							string newstring = currentstring;
							string tempstring1;
							string tempstring2;
							tempstring1 = newstring.substr(0, index);
							if (index == 0)
							{
								tempstring1 = "";
							}
							newstring = currentstring;
							if (index + rule.first.length() > newstring.length())
							{
								tempstring2 = "";
							}
							else
							{
								tempstring2 = newstring.substr(index + rule.first.length());
							}
							newstring = tempstring1 + rule.second[0] + tempstring2;
							//cout << currentstring << " " << newstring << " " << index << " " << rule.second[1] << " " << iteration + 1<< "\n";
							//memoize[currentstring].push_back(newstring);
							//memoizeparams[currentstring][newstring] = {};
							//memoizeparams[currentstring][newstring].push_back(rule.second[1]);
							//memoizeparams[currentstring][newstring].push_back(to_string(index + 1));
							vector<string> arg1 = rulesnum;
							vector<int> arg2 = indexes;
							vector<string> arg3 = past;
							arg1.push_back(rule.second[1]);
							arg2.push_back(index + 1);
							arg3.push_back(newstring);
							bool out = loop(newstring, endstring, arg1, arg2, arg3, iteration + 1, maxiteration, rules);
							if (out)
							{
								return true;
							}
						}
					}
				}
			}
			else
			{
				/*
				for (string next: memoize[currentstring])
				{
					vector<string> arg1 = rulesnum;
					vector<int> arg2 = indexes;
					vector<string> arg3 = past;
					arg1.push_back(memoizeparams[currentstring][next][0]);
					arg2.push_back(stoi(memoizeparams[currentstring][next][1]));
					arg3.push_back(next);
					bool out = loop(next, endstring, arg1, arg2, arg3, iteration + 1, maxiteration, rules);
					if (out)
					{
						return true;
					}
				}
				*/
				
			}
			
		}
	}
	return false;
	
}

int main()
{
    unordered_map<string, vector<string>> rules;
	string temp1;
	string temp2;
	string start;
	string finish;
	int count;
	for (int i = 1; i < 4; i++)
	{
		cin >> temp1 >> temp2;
		vector<string> tempv_vector = { temp2, to_string(i) };
		rules[temp1] = tempv_vector;
	}
	cin >> count >> start >> finish;
	loop(start, finish, {}, {}, {}, 0, count, rules);
}

File: 2019/J5.py
class RollingHash:
    def __init__(self, text, sizeWord):
        self.text = text
        self.hash = 0
        self.sizeWord = sizeWord

        for i in range(0, sizeWord):
            self.hash += (ord(self.text[i]) - ord("a")+1)*(2**(sizeWord - i -1))

        self.window_start = 0
        self.window_end = sizeWord

    def move_window(self):
        if self.window_end <= len(self.text) - 1:
            self.hash -= (ord(self.text[self.window_start]) - ord("a")+1)*2**(self.sizeWord-1)
            self.hash *= 2
            self.hash += ord(self.text[self.window_end])- ord("a")+1
            self.window_start += 1
            self.window_end += 1

    def window_text(self):
        return self.text[self.window_start:self.window_end]


def rabin_karp(word, text):
    if word == "" or text == "":
        return []
    if len(word) > len(text):
        return []
    returned = []
    rolling_hash = RollingHash(text, len(word))
    word_hash = RollingHash(word, len(word))

    for i in range(len(text) - len(word) + 1):
        if rolling_hash.hash == word_hash.hash:
            if rolling_hash.window_text() == word:
                returned.append(i)
        rolling_hash.move_window()
    return returned


sub1 = input().split()
sub2 = input().split()
sub3 = input().split()
s, i, f = input().split()
s = int(s)

rules = {sub1[0]: [sub1[1], 1, len(sub1[0]), len(sub1[1])], sub2[0]: [sub2[1], 2, len(sub2[0]), len(sub2[1])], sub3[0]: [sub3[1], 3, len(sub3[0]), len(sub3[1])]}
seen = {}
done = False


def loop(iteration, rtypes, indexes, past, string):
    global done
    if not done:
        if iteration == s and string == f:
            for index in range(s):
                print(rtypes[index], indexes[index], past[index])
                done = True
            return True
        if (string, iteration) in seen:
            return False
        elif iteration == s:
            return False
        else:
            seen[(string, iteration)] = True
            for rule in rules:
                found = rabin_karp(rule, string)
                for index in found:
                    newstring = string[:index] + rules[rule][0] + string[index:]
                    newstring = newstring[:index + rules[rule][3]] + newstring[index + rules[rule][3] + rules[rule][2]:]
                    arg1 = rtypes + [rules[rule][1]]
                    arg2 = indexes + [index + 1]
                    arg3 = past + [newstring]
                    out = loop(iteration + 1, arg1, arg2, arg3, newstring)
                    if out:
                        return True

    return False


loop(0, [], [], [], i)




